core/src/main/java/org/elasticsearch/plugins/Plugin.java

```diff
@@ -123,6 +125,17 @@ public void onIndexModule(IndexModule indexModule) {}
     public List<String> getSettingsFilter() { return Collections.emptyList(); }
 
     /**
+     * Provides a function to modify the metadata on startup.
+     * <p>
+     * Plugins should return the input metadata via {@link Function#identity()} if no upgrade is required.
+     * @return Never {@code null}. The same or upgraded {@code MetaData}.
+     * @throws IllegalStateException if the node needs to be stopped because {@code MetaData} is unsupported
```

Comment:
I would say something like "the node should not start" instead of the part about stopping (since this is during startup, while stopping implies the node is already started).